### ISSUE TITLE
Keras out of GPU memory

### ISSUE THREAD

**Issue by brianprichardson** [suragnair/alpha-zero-general#26]
Othello Keras after 34 iterations was out of gpu memory.
I have an 8GB GTX1070 but limited to per_process_gpu_memory_fraction = 0.4 (about 3.2GB)

Of course, I can run with more, but perhaps there should be some gpu memory size guidelines in the readme, assuming it is not an error.

Caused by op 'batch_normalization_199/FusedBatchNorm', defined at:
  File "main.py", line 29, in <module>
    c.learn()
  File "/home/brian/hitme/bin/alpha-zero-general/Coach.py", line 90, in learn
    pnet = self.nnet.__class__(self.game)
  File "/home/brian/hitme/bin/alpha-zero-general/othello/keras/NNet.py", line 27, in __init__
    self.nnet = onnet(game, args)
  File "/home/brian/hitme/bin/alpha-zero-general/othello/keras/OthelloNNet.py", line 37, in __init__
    h_conv1 = Activation('relu')(BatchNormalization(axis=3)(Conv2D(args.num_channels, 3, padding='same')(x_image)))         # batch_size  x board_x x board_y x num_channels
  File "/home/brian/hitme/lib/python3.6/site-packages/keras/engine/topology.py", line 617, in __call__
    output = self.call(inputs, **kwargs)
  File "/home/brian/hitme/lib/python3.6/site-packages/keras/layers/normalization.py", line 181, in call
    epsilon=self.epsilon)
  File "/home/brian/hitme/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 1824, in normalize_batch_in_training
    epsilon=epsilon)
  File "/home/brian/hitme/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py", line 1799, in _fused_normalize_batch_in_training
    data_format=tf_data_format)
  File "/home/brian/hitme/lib/python3.6/site-packages/tensorflow/python/ops/nn_impl.py", line 831, in fused_batch_norm
    name=name)
  File "/home/brian/hitme/lib/python3.6/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 2034, in _fused_batch_norm
    is_training=is_training, name=name)
  File "/home/brian/hitme/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/brian/hitme/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2956, in create_op
    op_def=op_def)
  File "/home/brian/hitme/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1470, in __init__
    self._traceback = self._graph._extract_stack()  # pylint: disable=protected-access

ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[64,512,6,6]
	 [[Node: batch_normalization_199/FusedBatchNorm = FusedBatchNorm[T=DT_FLOAT, data_format="NHWC", epsilon=0.001, is_training=true, _device="/job:localhost/replica:0/task:0/device:GPU:0"](conv2d_133/BiasAdd, batch_normalization_199/gamma/read, batch_normalization_199/beta/read, batch_normalization_199/Const_4, batch_normalization_199/Const_4)]]
	 [[Node: loss_33/add/_17985 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_3237_loss_33/add", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]

**Comment by evg-tyurin** [comment]
When Keras+TF is used you can try K.clear_session() after every Arena.playGame loop. However, if you clear_session() you should then recreate NNet object from the correct checkpoint.

**Comment by suragnair** [comment]
@evg-tyurin would it be possible to add K.clear_session() to the `load_checkpoint` function?

**Comment by evg-tyurin** [comment]
I think not. In that case the logic of Coach would be more complex and less readable. Besides, in that case you should recreate nnet too when pnet.load_checkpoint().

I'd propose (I've already implemented and tested it) to clear_session() at the end of each iteration when Arena is completed.
Coach.py
            `if <condition_to_accept_model>:
                print('REJECTING NEW MODEL')
                #self.nnet = pnet
                **self.nnet.recreate(folder=self.args.checkpoint, filename='temp.pth.tar')**
            else:
                print('ACCEPTING NEW MODEL')
                self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='checkpoint_' + str(i) + '.pth.tar')
                self.nnet.save_checkpoint(folder=self.args.checkpoint, filename='best.pth.tar')
                **self.nnet.recreate(folder=self.args.checkpoint, filename='best.pth.tar')**`
**new method** in ALL game/keras/NNet.py
    def recreate(self, folder='checkpoint', filename='checkpoint.pth.tar'):
        self.nnet.destroy()
        self.nnet = onnet(self.game, args)
        self.load_checkpoint(folder, filename)

**new method** in all game/keras/GameNNet.py
    def destroy(self):
        K.clear_session()

The best way is to try to move new methods from game specific classes to the base classes.

**Comment by suragnair** [comment]
Is the root cause the fact that unreferenced graphs are not deleted automatically?

**Comment by evg-tyurin** [comment]
Maybe, I'm not so handy with TF internals.

**Comment by suragnair** [comment]
Could you make a pull request for this? I believe we can do identical things for TensorFlow.

**Comment by evg-tyurin** [comment]
Of course, see #30

**Comment by suragnair** [comment]
f391999 should solve this. Can you do a run and let me know? You can start with a smaller `per_process_gpu_memory_fraction` if you wish.

**Comment by suragnair** [comment]
@brianprichardson any updates?

**Comment by brianprichardson** [comment]
Doing a run now.

**Comment by brianprichardson** [comment]
Up to 30 iterations so far with no sign of gpu memory issues.  I'll leave it running overnight.

**Comment by brianprichardson** [comment]
Ran for 154 generations with no gpu memory issues.

Uploading best weights as pull request.
